Serialization optimizations

AndrasKovacs opened the pull request:
Cut down total serialization + deserialization time for the stdlib benchmark by around 30%. Build times are also improved for serialization modules.

- Fixing strictness bugs. The `icodeN` function previously always failed to specialize. Making it properly specialize yields most of the speedups here.
- Moving `tickICode` logic out of line.
- Removing `ExceptT TypeError` from deserialization and using native `ErrorCall` exceptions instead. Nothing in the existing code actually relied on the richer type information in `TypeError`, nor on `ExceptT`. 
- Using unboxed `Int32` lists as `Node` instead of `[Int32]`. This reduces `Node` heap size by 40%.  
- Using simple association lists instead of `HashTable` in `Memo` in deserialization. These can never get large so it makes sense to use the lightest data representation.

AndrasKovacs committed:
Strictness optimizations on (de)serialization

AndrasKovacs committed:
Unbox Int32 in serialization nodes

AndrasKovacs committed:
Use exceptions instead of ExceptT in deserialization

AndrasKovacs committed:
Streamline TypeRep indexing in deserialization

Use a plain unboxed association list instead of HashTable-s. Type collisions
are rare and we should not ever have many collisions for the same node of
serialized data (since it's bounded by the number of different serialized
types in the absolute worst case).

andreasabel commented:
@AndrasKovacs Do these commits build individually and should be preserved, or should they be squashed?
Please remove one of the labels `pr: preserve commits` and `pr: squash-me` accordingly (I set both now to alert you of their presense).

andreasabel commented on src/full/Agda/TypeChecking/Serialise/Base.hs:
```diff
@@ -60,7 +60,6 @@ farEmpty = FreshAndReuse 0
 #ifdef DEBUG
                            0
 #endif
-
```
Removing this line looks like an accident.

andreasabel commented on src/full/Agda/TypeChecking/Serialise/Base.hs:
```diff
@@ -174,6 +173,7 @@ type R a = ExceptT TypeError (StateT St IO) a
 
 malformed :: R a
 malformed = throwError $ GenericError "Malformed input."
+{-# noinline malformed #-}
```
Could you please comment on the "why" of this NOINLINE?

Intuitively, one would think such a short piece of code should be inlined.

AndrasKovacs replied:
Cold/impossible code should be NOINLINE, generally speaking, to avoid pollution of hot code and minimize compile times.

andreasabel commented on src/full/Agda/TypeChecking/Serialise/Base.hs:
```diff
@@ -190,20 +190,25 @@ class Typeable a => EmbPrj a where
+  liftIO $ do
+    n <- fromMaybe 0 <$> H.lookup hmap key
+    H.insert hmap key $! n + 1
+{-# noinline goTickIcode #-}
```
Could you please comment on the "why" of this NOINLINE?

AndrasKovacs replied:
`tickICode` is again very cold code that only does something when the corresponding profiling flag is enabled, so this should be moved out of line (also to decrease compile times). Ideally, `tickICode` code could be toggled with a build flag. Even the simple branch on `whenM (asks collectStats)` adds about 10% code size to the Core, partly because GHC is happy to inline the same continuations in both branches.

andreasabel replied:
Great explanations!, great to have them also as comments in the source code.

andreasabel commented on src/full/Agda/TypeChecking/Serialise/Base.hs:
```diff
@@ -190,20 +190,25 @@ class Typeable a => EmbPrj a where
-      n <- fromMaybe 0 <$> H.lookup hmap key
-      H.insert hmap key $! n + 1
+tickICode _ = whenM (asks collectStats) $ goTickIcode (Proxy :: Proxy a)
+{-# inline tickICode #-}
```
The current style is UPPERCASE for GHC pragmas.

andreasabel replied:
Not sure we have to enforce this style, though...

andreasabel commented on src/full/Agda/TypeChecking/Serialise/Instances/Common.hs:
```diff
@@ -467,7 +468,7 @@ instance EmbPrj OpaqueId where
 instance (Eq k, Hashable k, EmbPrj k, EmbPrj v) => EmbPrj (HashMap k v) where
   icod_ m = mapPairsIcode (HMap.toList m)
-  value = vcase (fmap HMap.fromList . mapPairsValue)
+  value = vcase ((<$!>) HMap.fromList . mapPairsValue)
```
Stylewise, I'd use a section.
```haskell
value = vcase $ (HMap.fromList <$!>) . mapPairsValue
```

AndrasKovacs commented:
Squashed. I removed the label.

andreasabel commented on src/full/Agda/TypeChecking/Serialise/Instances/Common.hs:
```diff
@@ -707,7 +708,7 @@ instance EmbPrj SrcLoc where
 instance EmbPrj CallStack where
   icod_ = icode . getCallStack
-  value = fmap fromCallSiteList . value
+  value = (<$!>) fromCallSiteList . value
```
ditto

andreasabel commented on src/full/Agda/TypeChecking/Serialise/Base.hs:
```diff
@@ -440,19 +510,19 @@ instance VALU t 'True where
   valueArgs _ xs = case xs of
-    (x : xs') -> (x,) <$> valueArgs (Proxy :: Proxy t) xs'
-    _         -> Nothing
+    x : xs' -> do {!node <- valueArgs (Proxy :: Proxy t) xs'; pure (Pair x node)}
```
Why cannot this be written with the strict `fmap`, `(<$!>)`?

Something like:
```haskell
Pair x <$!> valueArgs (Proxy :: Proxy t) xs'
```

AndrasKovacs replied:
No reason, it does look better that way.

andreasabel commented on src/full/Agda/TypeChecking/Serialise/Instances/Common.hs:
```diff
@@ -214,13 +215,13 @@ instance EmbPrj TopLevelModuleName where
+  icod_ xs = icodeNode =<< go xs where
+    go :: [a] -> S Node
+    go []     = pure Empty
+    go (a:as) = do {n <- icode a; ns <- go as; pure $! Cons n ns}
```
The manual inlining of `mapM icode` deserves a comment / justification.

AndrasKovacs replied:
We go from lists to `Node` so `mapM` isn't well-typed here.

andreasabel commented on src/full/Agda/TypeChecking/Serialise/Base.hs:
```diff
@@ -362,57 +428,61 @@ icodeMemo getDict getCounter a icodeP = do
 
 instance ICODE t (IsBase t) => ICODE (a -> t) 'False where
-  icodeArgs _ (a , as) = icode a >>= \ hd -> (hd :) <$> icodeArgs (Proxy :: Proxy t) as
+  icodeArgs _ (Pair a as) = icode a >>= \ !hd -> do
```
Please use `do`instead of `>>=` here.

andreasabel commented on src/full/Agda/TypeChecking/Serialise/Base.hs:
```diff
@@ -362,57 +428,61 @@ icodeMemo getDict getCounter a icodeP = do
+      -- read new value and save it
+      (do !v <- valu . (! ix) =<< gets nodeE
+          liftIO $ writeArray memo ix $! MECons fp (unsafeCoerce v) slot
+          return v)
```
Outermost parentheses not needed due to `LANGUAGE BlockArguments`.

andreasabel commented on src/full/Agda/TypeChecking/Serialise/Base.hs:
```diff
@@ -39,8 +45,65 @@ import Agda.Utils.Pointer
+
+instance B.Binary Node where
+
+  get = do {n :: Int <- B.get; go n} where
```
Wouldn't this work?:
```haskell
get = go =<< B.get
  where
  ...
```

andreasabel reviewed (changes requested):
Thanks for this great work @AndrasKovacs !

I have a few change request concerning style and documentation.

I see you added a lot of strictness annotations.  
Could you formulate a policy on when to use strict and when lazy in the serializer, a policy that other developers could follow so that your work here does not get destroyed by future contributions?

Also, I think that choices of NOINLINE and INLINE should get a comment in the source code justifying the choice.

AndrasKovacs committed:
Style fix, more comments, more strictness annotations

AndrasKovacs commented:
I addressed the conversations, added some new strictness annotations that I missed (which may or may not matter for overall performance), added some comments on pragmas and strictness, and added a [link](https://github.com/tkaitchuck/aHash/wiki/AHash-fallback-algorithm) to where the `Node` hash function is adapted from.

Also reran std-lib bench, the concrete data that I've got:
- Before: serialization 67.48s, deserialization 15.05s
- After: serialization 45.03s, deserialization 15.04s

There's quite a bit of noise in my laptop results that may have eaten the deserialization gains here, but the total improvement is clear.

andreasabel commented on src/full/Agda/TypeChecking/Serialise.hs:
```diff
@@ -207,24 +207,30 @@ decode s = do
 
-  (mf, r) <- liftIO $ E.handle (\(E.ErrorCall s) -> noResult s) $ do
-
+  res <- liftIO $ E.handle (\(E.ErrorCall s) -> pure Nothing) $ do
```
`s` is dropped here, needlessly it seems.
It seems like we could communicate the reason `s` when using `Left/Right` instead of `Nothing/Just` here.

AndrasKovacs replied:
I added `s` propagation and printing in the failure case just below. The original error printing was commented out to not pull `Pretty` dependency, but now the message is just a `String`.

andreasabel commented on src/full/Agda/TypeChecking/Serialise/Base.hs:
```diff
@@ -167,43 +245,52 @@ type S a = ReaderT Dict IO a
 
   default icod_ :: (Enum a, Bounded a) => a -> S Int32
-  icod_ = return . fromIntegral . fromEnum
+  icod_ x = return $! fromIntegral $! fromEnum x
```
Ah, this line is what I was looking for previously but couldn't find it any more.
This would be an opportunity to use a strict composition `(.!)` which we could place in `Agda.Utils.Function`.

andreasabel commented on src/full/Agda/TypeChecking/Serialise/Base.hs:
```diff
@@ -362,57 +449,62 @@ icodeMemo getDict getCounter a icodeP = do
+  icodeArgs _ (Pair a as) = do
+    !hd   <- icode a
+    !node <- icodeArgs (Proxy :: Proxy t) as
+    pure $ Cons hd node
```
I am trying to understand whether a `$!` would make a difference here.
Below you have it, but no strict bindings of `hd` and `node`.
Are both ways equivalent?

AndrasKovacs replied:
`($!)` doesn't make a diff here because `Cons hd node` is a redex-free constructor expression for which GHC never creates a thunk. Alternatively, we can omit bangs from `hd` and `node`, and use `($!)`. That works because the fields of `Cons` are strict, so if `Cons` is whnf-ed then `hd` and `node` must be also.

andreasabel commented on src/full/Agda/TypeChecking/Serialise/Instances/Common.hs:
```diff
@@ -214,13 +215,12 @@ instance EmbPrj TopLevelModuleName where
+  icod_ xs = icodeNode =<< go xs where
+    go :: [a] -> S Node
+    go []     = pure Empty
+    go (a:as) = do {n <- icode a; ns <- go as; pure $! Cons n ns}
```
Compare this with the stye above where `n` and `ns` are bound strictly but we would have `pure $ Cons n ns` (no strict application).
Are both equivalent?

AndrasKovacs replied:
They are, because `Cons` has strict fields. Btw, using `Strict` in these modules would be IMO fine, instead of getting bangs right (and there are many equivalent ways of doing so).

andreasabel commented on src/full/Agda/TypeChecking/Serialise/Instances/Highlighting.hs:
```diff
@@ -124,20 +124,20 @@ instance EmbPrj a => EmbPrj (RM.RangeMap a) where
+      !start <- icode start
+      !end <- icode end
+      !entry <- icode entry
+      convert (Cons start (Cons end (Cons entry ys))) xs
```
Type signature for `convert` would be good for the code reader.

andreasabel commented on src/full/Agda/TypeChecking/Serialise/Instances/Highlighting.hs:
```diff
@@ -124,20 +124,20 @@ instance EmbPrj a => EmbPrj (RM.RangeMap a) where
-      entry <- value entry
+      !start <- value start
+      !end <- value end
+      !entry <- value entry
```
Type signature for `convert` would be good for the code reader.

andreasabel commented on src/full/Agda/TypeChecking/Serialise/Instances/Common.hs:
```diff
@@ -295,8 +295,8 @@ instance EmbPrj RangeFile where
     case r of
-      Left err -> throwError $ findErrorToTypeError m err
-      Right f  -> return $ RangeFile (srcFilePath f) (Just m)
+      Left err -> liftIO $ E.throwIO $ E.ErrorCall $ "file not found: " ++ show m
```
Can the `err` be appended here as additional detail?

AndrasKovacs replied:
Indeed, I added `FindError` and `SourceFile` `Show` instances, and used that to put `err` in the error.

andreasabel commented on src/full/Agda/Utils/TypeLevel.hs:
```diff
@@ -72,6 +72,17 @@ type family Constant (b :: l) (as :: [k]) :: [l] where
+
+strictCurry :: (StrictPair a b -> c) -> (a -> b -> c)
+strictCurry f = \ !a !b -> f (Pair a b)
+{-# inline strictCurry #-}
```
Since you did it for the other files already, please also uppercase INLINE in this file.

andreasabel reviewed (commented):
Thanks for the update!

I have a few new comments, please have a look.

AndrasKovacs committed:
Fixes, printing more error info on deserialization failure

andreasabel reviewed (approved):
Thanks again!

andreasabel commented:
This was merged as commit no 22.222:

<img width="835" alt="Screenshot 2023-10-07 at 16 10 14" src="https://github.com/agda/agda/assets/1155218/039fdd61-4a13-49a4-b998-d4d7db16ab2f">
